File: SortingAlgorithm/_9_BucketSort/BucketSort.py
def insertionSort(bucket):
    for i in range(1, len(bucket)):
        key = bucket[i]
        j = i - 1
        while j >= 0 and bucket[j] > key:
            bucket[j + 1] = bucket[j]
            j -= 1
        bucket[j + 1] = key
    return bucket


def bucketSort(arr):
    # Determine the number of buckets (slots)
    num_buckets = 10
    buckets = [[] for _ in range(num_buckets)]

    # Distribute elements into buckets
    for value in arr:
        index = int(value * num_buckets)
        buckets[index].append(value)

    # Sort individual buckets using Insertion Sort
    for i in range(num_buckets):
        buckets[i] = insertionSort(buckets[i])

    # Concatenate sorted buckets to get the final sorted array
    sorted_arr = [element for bucket in buckets for element in bucket]

    return sorted_arr


# Driver code
if __name__ == "__main__":
    x = [0.897, 0.565, 0.656, 0.1234, 0.665, 0.3434]
    sorted_x = bucketSort(x)
    print("Sorted Array:", sorted_x)

File: SortingAlgorithm/_9_BucketSort/BucketSort.java
package Algorithm.SortingAlgorithm._9_BucketSort;

import java.util.*;

public class BucketSort {
    public static <T extends Comparable<T>> void bucketSort(T[] arr) {
        int n = arr.length;

        if (n <= 0) {
            return;
        }

        // Create n empty buckets
        List<List<T>> buckets = new ArrayList<>(n);
        for (int i = 0; i < n; i++) {
            buckets.add(new ArrayList<>());
        }

        // Put array elements in different buckets
        for (int i = 0; i < n; i++) {
            int idx = (int) (n * ((Double) arr[i]).doubleValue());
            buckets.get(idx).add(arr[i]);
        }

        // Sort individual buckets
        for (int i = 0; i < n; i++) {
            Collections.sort(buckets.get(i));
        }

        // Concatenate all buckets into arr[]
        int index = 0;
        for (int i = 0; i < n; i++) {
            for (T element : buckets.get(i)) {
                arr[index++] = element;
            }
        }
    }

    public static void main(String[] args) {
        Double[] arr = {0.897, 0.565, 0.656, 0.1234, 0.665, 0.3434};

        bucketSort(arr);

        System.out.println("Sorted array is ");
        for (double el : arr) {
            System.out.print(el + " ");
        }
    }
}
